internal/kojiapi/server.go

```diff
@@ -23,66 +23,70 @@ import (
 
 // Server represents the state of the koji Server
 type Server struct {
+	echo        *echo.Echo
 	workers     *worker.Server
 	rpmMetadata rpmmd.RPMMD
 	distros     *distro.Registry
 	kojiServers map[string]koji.GSSAPICredentials
 }
 
 // NewServer creates a new koji server
-func NewServer(workers *worker.Server, rpmMetadata rpmmd.RPMMD, distros *distro.Registry, kojiServers map[string]koji.GSSAPICredentials) *Server {
-	server := &Server{
+func NewServer(logger *log.Logger, workers *worker.Server, rpmMetadata rpmmd.RPMMD, distros *distro.Registry, kojiServers map[string]koji.GSSAPICredentials) *Server {
+	s := &Server{
 		workers:     workers,
 		rpmMetadata: rpmMetadata,
 		distros:     distros,
 		kojiServers: kojiServers,
 	}
-	return server
+
+	s.echo = echo.New()
+	s.echo.Binder = binder{}
+	s.echo.StdLogger = logger
+
+	api.RegisterHandlers(s.echo, &apiHandlers{s})
+
+	return s
 }
 
 // Serve serves the koji API over the provided listener socket
-func (server *Server) Serve(listener net.Listener) error {
-	s := http.Server{Handler: api.Handler(server)}
+func (s *Server) Serve(listener net.Listener) error {
+	s.echo.Listener = listener
 
-	err := s.Serve(listener)
+	err := s.echo.Start("")
 	if err != nil && err != http.ErrServerClosed {
 		return err
 	}
 
 	return nil
 }
 
-// PostCompose handles a new /compose POST request
-func (server *Server) PostCompose(w http.ResponseWriter, r *http.Request) {
-	contentType := r.Header["Content-Type"]
-	if len(contentType) != 1 || contentType[0] != "application/json" {
-		http.Error(w, "Only 'application/json' content type is supported", http.StatusUnsupportedMediaType)
-		return
-	}
+// apiHandlers implements api.ServerInterface - the http api route handlers
+// generated from api/openapi.yml. This is a separate object, because these
+// handlers should not be exposed on the `Server` object.
+type apiHandlers struct {
+	server *Server
+}
 
+// PostCompose handles a new /compose POST request
+func (h *apiHandlers) PostCompose(ctx echo.Context) error {
 	var request api.ComposeRequest
-	err := json.NewDecoder(r.Body).Decode(&request)
+	err := ctx.Bind(&request)
 	if err != nil {
-		http.Error(w, "Could not parse JSON body", http.StatusBadRequest)
-		return
+		return err
 	}
 
-	d := server.distros.GetDistro(request.Distribution)
+	d := h.server.distros.GetDistro(request.Distribution)
 	if d == nil {
-		http.Error(w, fmt.Sprintf("Unsupported distribution: %s", request.Distribution), http.StatusBadRequest)
-		return
+		return echo.NewHTTPError(http.StatusBadRequest, "Unsupported distribution: %s", request.Distribution)
```

Comment:
Unfortunately, this function does not take format strings (cf [echo.NewHTTPError](https://pkg.go.dev/github.com/labstack/echo/v4#NewHTTPError). I realized this too late myself when working on the worker API and fixed it in https://github.com/osbuild/osbuild-composer/commit/ccebfb014c4bb53870d1c876ab4868c936d5178a.